From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Otto=20Pfl=C3=BCger?= <otto.pflueger@abscue.de>
Date: Sat, 17 May 2025 14:21:45 +0200
Subject: [PATCH 050/225] clk: sprd: add runtime PM support
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Some clock configuration registers can only be configured when a power
domain or gate clock is enabled. Add support for runtime power
management to allow using these registers without keeping the power
domain permanently enabled.

Signed-off-by: Otto Pflüger <otto.pflueger@abscue.de>
---
 drivers/clk/sprd/common.c | 39 +++++++++++++++++++++++++++++++++++++++
 drivers/clk/sprd/common.h |  9 +++++++++
 2 files changed, 48 insertions(+)

diff --git a/drivers/clk/sprd/common.c b/drivers/clk/sprd/common.c
index 0f2181617..cd4ed3a7a 100644
--- a/drivers/clk/sprd/common.c
+++ b/drivers/clk/sprd/common.c
@@ -5,10 +5,12 @@
 // Copyright (C) 2017 Spreadtrum, Inc.
 // Author: Chunyan Zhang <chunyan.zhang@spreadtrum.com>
 
+#include <linux/clk.h>
 #include <linux/mfd/syscon.h>
 #include <linux/module.h>
 #include <linux/of_address.h>
 #include <linux/platform_device.h>
+#include <linux/pm_runtime.h>
 #include <linux/regmap.h>
 
 #include "common.h"
@@ -41,10 +43,13 @@ int sprd_clk_regmap_init(struct platform_device *pdev,
 	void __iomem *base;
 	struct device *dev = &pdev->dev;
 	struct device_node *node = dev->of_node, *np;
+	struct sprd_clk_drvdata *data;
 	struct regmap *regmap;
 	struct resource *res;
 	struct regmap_config reg_config = sprdclk_regmap_config;
 
+	data = devm_kzalloc(&pdev->dev, sizeof(*data), GFP_KERNEL);
+
 	if (of_property_present(node, "sprd,syscon")) {
 		regmap = syscon_regmap_lookup_by_phandle(node, "sprd,syscon");
 		if (IS_ERR(regmap)) {
@@ -76,6 +81,14 @@ int sprd_clk_regmap_init(struct platform_device *pdev,
 
 	sprd_clk_set_regmap(desc, regmap);
 
+	data->gate_clk = devm_clk_get_optional_enabled(&pdev->dev, "enable");
+	if (IS_ERR(data->gate_clk))
+		return dev_err_probe(dev, PTR_ERR(data->gate_clk),
+				     "failed to get register gate clock\n");
+
+	data->regmap = regmap;
+	platform_set_drvdata(pdev, data);
+
 	return 0;
 }
 EXPORT_SYMBOL_GPL(sprd_clk_regmap_init);
@@ -109,5 +122,31 @@ int sprd_clk_probe(struct device *dev, struct clk_hw_onecell_data *clkhw)
 }
 EXPORT_SYMBOL_GPL(sprd_clk_probe);
 
+static int __maybe_unused sprd_clk_suspend(struct device *dev)
+{
+	struct sprd_clk_drvdata *data = dev_get_drvdata(dev);
+
+	clk_disable_unprepare(data->gate_clk);
+
+	return 0;
+}
+
+static int __maybe_unused sprd_clk_resume(struct device *dev)
+{
+	struct sprd_clk_drvdata *data = dev_get_drvdata(dev);
+	int ret;
+
+	ret = clk_prepare_enable(data->gate_clk);
+	if (ret) {
+		dev_err(dev, "failed to enable register gate clock: %d\n", ret);
+		return ret;
+	}
+
+	return 0;
+}
+
+EXPORT_GPL_RUNTIME_DEV_PM_OPS(sprd_clk_pm_ops, sprd_clk_suspend,
+			      sprd_clk_resume, NULL);
+
 MODULE_DESCRIPTION("Spreadtrum clock infrastructure");
 MODULE_LICENSE("GPL v2");
diff --git a/drivers/clk/sprd/common.h b/drivers/clk/sprd/common.h
index 1d077b39c..1f348bdef 100644
--- a/drivers/clk/sprd/common.h
+++ b/drivers/clk/sprd/common.h
@@ -24,6 +24,13 @@ struct sprd_clk_desc {
 	struct sprd_clk_common		**clk_clks;
 	unsigned long			num_clk_clks;
 	struct clk_hw_onecell_data      *hw_clks;
+	bool				need_regcache;
+};
+
+struct sprd_clk_drvdata {
+	struct regmap	*regmap;
+	struct clk	*gate_clk;
+	bool		cache_enabled;
 };
 
 static inline struct sprd_clk_common *
@@ -35,4 +42,6 @@ int sprd_clk_regmap_init(struct platform_device *pdev,
 			 const struct sprd_clk_desc *desc);
 int sprd_clk_probe(struct device *dev, struct clk_hw_onecell_data *clkhw);
 
+extern const struct dev_pm_ops sprd_clk_pm_ops;
+
 #endif /* _SPRD_CLK_COMMON_H_ */
